frontend/src/components/widgets/DataFrame/hooks/useDataLoader.ts

```diff
@@ -0,0 +1,304 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react"
+
+import { GridCell, DataEditorProps } from "@glideapps/glide-data-grid"
+
+import { Quiver } from "src/lib/Quiver"
+import { Arrow as ArrowProto } from "src/autogen/proto"
+import { notNullOrUndefined, isNullOrUndefined } from "src/lib/utils"
+import { logWarning, logError } from "src/lib/log"
+
+import {
+  getColumnTypeFromArrow,
+  getColumnsFromArrow,
+  getCellFromArrow,
+} from "src/components/widgets/DataFrame/arrowUtils"
+import EditingState from "src/components/widgets/DataFrame/EditingState"
+import {
+  BaseColumn,
+  BaseColumnProps,
+  getErrorCell,
+  ColumnTypes,
+  ColumnCreator,
+} from "src/components/widgets/DataFrame/columns"
+
+// Using this ID for column config will apply the config to all index columns
+export const INDEX_IDENTIFIER = "index"
+// Prefix used in the config column mapping when referring to a column via the numeric position
+export const COLUMN_POSITION_PREFIX = "col:"
+
+/**
+ * Options to configure columns.
+ */
+export interface ColumnConfigProps {
+  width?: number
+  title?: string
+  type?: string
+  hidden?: boolean
+  editable?: boolean
+  metadata?: Record<string, unknown>
+  alignment?: string
+}
+
+/**
+ * Apply the user-defined column configuration if supplied.
+ *
+ * @param columnProps - The column properties to apply the config to.
+ * @param columnsConfig - The user-defined column configuration.
+ *
+ * @return the column properties with the config applied.
+ */
+export function applyColumnConfig(
+  columnProps: BaseColumnProps,
+  columnsConfig: Map<string | number, ColumnConfigProps>
+): BaseColumnProps {
+  if (!columnsConfig) {
+    // No column config configured
+    return columnProps
+  }
+
+  let columnConfig
```

Comment:
Can we change the naming on this? Maybe to `singleColumnConfig`? 
`columnConfig` and `columnSConfig` was super difficult to read and I thought line 70-73 and line 91-94 were the same code 😆 

Also just to double check, it looks like this is applying some configs to each column?

Reply:
Yes, good point. I changed `columnsConfig` to `columnConfigMapping` since it is a mapping between a specific column and it's column configuration properties. `columnConfig` is in this case just the properties of a single column.